### PR TITLE
Added ability to check if no message exists as a test to store the message.

### LINKED ISSUE #2494: Ability to publish to a kv style topic only if empty
#### Use Case:

Today we have `Nats-Expected-Last-Subject-Sequence` that limits publishing to a stream if the previous message had a specific sequence

Along the same lines it would be useful to ensure only the first client can write a message when the subject is empty.

We discussed maybe (ab)using `Nats-Expected-Last-Subject-Sequence:0` but decided not to, so maybe `Nats-Expected-Empty-Subject:1`

A sample use case could be for example registering a schema that relates to a specific topic, the first publisher to this topic should register the subject but that's true only ever once.

This setting should probably only be accepted in `max_msgs_per_subject > 0` stream

### PR DESCRIPTION
**Pull request by derekcollison**
Feature request.

Signed-off-by: Derek Collison <derek@nats.io>

Resolves #2494

/cc @nats-io/core

### CODE REVIEW

File: server/jetstream_test.go [hunk 1]
```diff
@@ -12973,6 +12973,46 @@ func TestJetStreamPerSubjectPending(t *testing.T) {
...
+
+	// This should succeed.
+	m := nats.NewMsg("KV.33")
+	m.Header.Set(JSExpectedLastSubjSeq, "0")
+	if _, err := js.PublishMsg(m); err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	// This should fail.
+	m = nats.NewMsg("KV.22")
```
Comment (ripienaar) [inline server/jetstream_test.go:13009]: Does it again work after purging the subject?

**Review by ripienaar — COMMENTED** [review]
Question about behaviour

File: server/jetstream_test.go (hunk 1 above)
Comment (derekcollison) [reply server/jetstream_test.go:13009]: Yes, just updated the test as well.

**Review by ripienaar — APPROVED** [review]
LGTM

File: server/stream.go [hunk 2]
```diff
@@ -2686,9 +2686,13 @@ func (mset *stream) processJetStreamMsg(subject, reply string, hdr, msg []byte,
 			return fmt.Errorf("last msgid mismatch: %q vs %q", lmsgId, last)
 		}
 		// Expected last sequence per subject.
-		if seq := getExpectedLastSeqPerSubject(hdr); seq > 0 {
+		if seq, exists := getExpectedLastSeqPerSubject(hdr); exists {
 			// TODO(dlc) - We could make a new store func that does this all in one.
 			_, lseq, _, _, _, err := mset.store.LoadLastMsg(subject)
+			// If seq passed in is zero that signals we expect no msg to be present.
```
Comment (kozlovic) [inline server/stream.go:2692]: I think that this change may break MQTT code that uses this header:
```
=== RUN   TestMQTTCluster
=== RUN   TestMQTTCluster/first_start
=== RUN   TestMQTTCluster/first_start/qos_0
    mqtt_test.go:2755: Error reading: EOF
--- FAIL: TestMQTTCluster (7.10s)
    --- FAIL: TestMQTTCluster/first_start (6.68s)
        --- FAIL: TestMQTTCluster/first_start/qos_0 (1.13s)
```
I set this header and initially seq is 0, however, it does not mean that I was not expecting any message to be set. Of course, we can fix the MQTT code and not set the header if know seq is 0.

**Review by kozlovic — CHANGES_REQUESTED** [review]

**Comment by ripienaar** [comment]
Yeah looks good to me apart from what Ivan says

File: server/stream.go (hunk 2 above)
Comment (kozlovic) [reply server/stream.go:2692]: Actually, no, I don't think MQTT code should change. If I don't set the expected to 0, then I have the issue where 2 or more may be updating at the same time, and this is wrong.

So I think overloading the meaning of this header should be rethought.

Comment (kozlovic) [reply server/stream.go:2692]: Humm... I guess original code would have ignored it anyway, so for sure we could fix the MQTT code and/or mqtt test..

Comment (derekcollison) [reply server/stream.go:2692]: We paired and fixed.

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**